SwiftIfConfig: A library to evaluate `#if` conditionals within a Swift syntax tree.

DougGregor opened the pull request:
Swift provides the ability to conditionally compile parts of a source file based on various built-time conditions, including information about the target (operating system, processor architecture, environment), information about the compiler (version, supported attributes and features), and user-supplied conditions specified as part of the build (e.g., `DEBUG`), which we collectively refer to as the *build configuration*. These conditions can occur within a `#if` in the source code, e.g.,

```swift
func f() {
#if DEBUG
  log("called f")
#endif

#if os(Linux)
  // use Linux API
#elseif os(iOS) || os(macOS)
  // use iOS/macOS API
#else
  #error("unsupported platform")
#endif
}
```

The syntax tree and its parser do not reason about the build configuration. Rather, the syntax tree produced by parsing this code will include `IfConfigDeclSyntax` nodes wherever there is a `#if`, and each such node contains the a list of clauses, each with a condition to check (e.g., `os(Linux)`) and a list of syntax nodes that are conditionally part of the program. Therefore, the syntax tree captures all the information needed to process the source file for any build configuration.

The `SwiftIfConfig` library provides utilities to determine which syntax nodes are part of a particular build configuration. Each utility requires that one provide a specific build configuration (i.e., an instance of a type that conforms to the <doc:BuildConfiguration> protocol), and provides a different view on essentially the same information:

* <doc:ActiveSyntaxVisitor> and <doc:ActiveSyntaxAnyVisitor> are visitor types that only visit the syntax nodes that are included ("active") for a given build configuration, implicitly skipping any nodes within inactive `#if` clauses.
* `SyntaxProtocol.removingInactive(in:)` produces a syntax node that removes all inactive regions (and their corresponding `IfConfigDeclSyntax` nodes) from the given syntax tree, returning a new tree that is free of `#if` conditions.
* `IfConfigDeclSyntax.activeClause(in:)` determines which of the clauses of an `#if` is active for the given build configuration, returning the active clause.
* `SyntaxProtocol.isActive(in:)` determines whether the given syntax node is active for the given build configuration.

DougGregor commented:
@swift-ci please test

harlanhaskins commented:
I would love to use this in the part of the compiler that strips inactive `#if` conditions out of module interfaces

ahoppen commented on Tests/SwiftIfConfigTest/EvaluateTests.swift:
```diff
@@ -0,0 +1,149 @@
+      try IfConfigState(condition: condition, configuration: buildConfig)
+    }
+
+    XCTAssertThrowsError(try ifConfigState("3.14159")) { error in
+      XCTAssertEqual(String(describing: error), "invalid conditional compilation expression")
+    }
```
If `#if` condition evaluation produces diagnostics, it would probably be good to have a `assertIfConfigEvaluation` that accepts a `DiagnosticSpec` and location markers.

And honestly, I think it would be a good idea even in the current design to avoid the repeated definition of `ifConfigState`.

DougGregor replied:
This is done now

ahoppen commented on Tests/SwiftIfConfigTest/VisitorTests.swift:
```diff
@@ -0,0 +1,183 @@
+    super.init(viewMode: .sourceAccurate, configuration: configuration)
+  }
+  open override func visitAny(_ node: Syntax) -> SyntaxVisitorContinueKind {
+    XCTAssertTrue(try! node.isActive(in: configuration))
```
Want to wrap this in `XCTAssertNoThrows` instead of `try!`?

DougGregor replied:
Sure

DougGregor commented on Sources/SwiftIfConfig/BuildConfiguration.swift:
```diff
@@ -0,0 +1,92 @@
+  /// Custom build conditions can be set by the `-D` command line option to
+  /// the Swift compiler. For example, `-DDEBUG` sets the custom condition
+  /// named `DEBUG`.
+  func isCustomConditionSet(name: String, syntax: ExprSyntax) -> Bool?
```
Done!

DougGregor commented on Sources/SwiftIfConfig/IfConfigFunctions.swift:
```diff
@@ -0,0 +1,53 @@
+//===----------------------------------------------------------------------===//
+
+/// Enum capturing all of the functions that can be used in an `#if` condition.
+enum IfConfigFunctions: String {
```
Done!

DougGregor commented on Sources/SwiftIfConfig/SyntaxLiteralUtils.swift:
```diff
@@ -0,0 +1,48 @@
+extension TupleExprSyntax {
+  /// Whether this tuple is a parenthesized expression, e.g., (x).
+  var isParentheses: Bool {
+    guard elements.count == 1, let element = elements.first else { return false }
+    return element.label == nil
```
Yes, you're right of course. I wrote this first and didn't come back to it.

DougGregor commented on Sources/SwiftIfConfig/IfConfigState.swift:
```diff
@@ -0,0 +1,24 @@
+
+/// Describes the state of a particular region guarded by `#if` or similar.
+public enum IfConfigState {
+  /// The region is not parsed, and may contain syntax that is invalid.
+  ///
+  /// TODO: For now, the IfConfig library does not distinguish between
+  /// inactive and unparsed regions, so this case is never used.
+  case unparsed
```
Yes, the old parser does this. Essentially, if you have something that isn't like a compiler version check, the code in there needs to be syntactically valid, e.g., 
 
```swift
#if DEBUG
func mustBeSyntacticallValid() { }
#endif
```

but if you do a compiler check, an older compiler won't parse it at all

```swift
#if compiler(>=5.10)
this is not syntactically valid code, but Swift compilers prior to 5.10 won't care
#endif
```

For the new parser, this means we should be emitting parser diagnostics for both `active` and `inactive` regions, but not for `unparsed` regions.

DougGregor commented on Sources/SwiftIfConfig/SyntaxLiteralUtils.swift:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    // FIXME: Handle escaping here.
+    return identExpr.identifier.text
```
I'm fine with losing unicode round-tripness for invalid code; we're in the realm of dealing with well-formed code, and we should work with identifiers.

That said, I do think we need to come up with a standard way of turning `TokenSyntax` nodes into identifiers, so there's a central place where we deal with escaped keywords like ``case``, and we have one place to implement and document rules like what normalization happens. I don't think it's the role of this pull request to define that, though.

ahoppen replied:
Just an idea so we don’t need to plough through all of this code again in the search for identifiers.

What do you think about defining the following. That way we would at least have the abstraction layer and we can iterate on it.

```swift
struct Identifier {
  var name: String

  init(_ token: TokenSyntax) {
     // FIXME: Handle escaping
    self.name = token.text
  }
}

extension TokenSyntax {
  var identifier: Identifier {
    Identifier(self)
  }
}
```

DougGregor commented on Sources/SwiftIfConfig/VersionTuple.swift:
```diff
@@ -0,0 +1,58 @@
+
+extension VersionTuple {
+  /// Create a version tuple from its components.
+  public init(_ components: Int...) {
+    self.components = components
+  }
```
If we're going to have a version type, I think we need to have a way to create one that isn't parsing a string. So yes, I'd like to keep it.

ahoppen replied:
OK, SGTM.

DougGregor commented on Sources/SwiftIfConfig/VersionTuple.swift:
```diff
@@ -0,0 +1,58 @@
+public struct VersionTuple {
+  /// The components of the version tuple, start with the major version.
+  public var components: [Int]
+}
+
+extension VersionTuple {
```
Sure, I'll collapse them.

DougGregor commented on Sources/SwiftIfConfig/SwiftIfConfig.docc/SwiftIfConfig.md:
```diff
@@ -0,0 +1,34 @@
+The `SwiftIfConfig` library provides utilities to determine which syntax nodes are part of a particular build configuration. Each utility requires that one provide a specific build configuration (i.e., an instance of a type that conforms to the <doc:BuildConfiguration> protocol), and provides a different view on essentially the same information:
+
+* <doc:ActiveSyntaxVisitor> and <doc:ActiveSyntaxAnyVisitor> are visitor types that only visit the syntax nodes that are included ("active") for a given build configuration, implicitly skipping any nodes within inactive `#if` clauses.
+* `SyntaxProtocol.removingInactive(in:)` produces a syntax node that removes all inactive regions (and their corresponding `IfConfigDeclSyntax` nodes) from the given syntax tree, returning a new tree that is free of `#if` conditions.
```
I think so? I'll try it.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+
+  // Integer literals evaluate true if that are not "0".
+  if let intLiteral = condition.as(IntegerLiteralExprSyntax.self) {
+    return intLiteral.digits.text != "0"
```
Yeah, it's pretty weird, but I'm following what the C++ implementation does for now.

ahoppen replied:
OK, in that case: Whatever. A short comment that this is weird because it matches the C++ implementation would be nice IMO.

DougGregor commented on Sources/SwiftIfConfig/VersionTuple.swift:
```diff
@@ -0,0 +1,58 @@
+      }
+
+      components.append(component)
+    }
```
Good point, I'll do that.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+import SwiftSyntax
+import SwiftOperators
+
+enum IfConfigError: Error, CustomStringConvertible {
```
I definitely agree that we should be able to form `Diagnostic` instances. Given the need for warnings, I'm thinking about providing diagnostics (not errors), and if there will be a way for the APIs to throw, we can throw `DiagnosticsError`. We could do something similar to the approach of `SwiftOperators` and have a "diagnostic handler" that we stream out diagnostics to. It can be allowed to throw, and will make all of the APIs `rethrows` to account for cases where the client decides to throw.

But it would be good to settle on some kind of pattern here.

DougGregor commented on Tests/SwiftIfConfigTest/EvaluateTests.swift:
```diff
@@ -0,0 +1,149 @@
+    XCTAssertEqual(try ifConfigState("true"), .active)
+    XCTAssertEqual(try ifConfigState("false"), .inactive)
+
+    // FIXME: How can we produce warnings from this code?
```
Yup.

DougGregor commented on Tests/SwiftIfConfigTest/TestingBuildConfiguration.swift:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  func isActiveTargetArchitecture(name: String, syntax: SwiftSyntax.ExprSyntax) -> Bool? {
+    name == "arm64"
```
Because this build configuration is all I need for testing. See my TODO item about a "fixed" build configuration or similar; this can grow into that.

DougGregor commented on Tests/SwiftIfConfigTest/EvaluateTests.swift:
```diff
@@ -0,0 +1,149 @@
+  func testCanImport() throws {
+    let buildConfig = TestingBuildConfiguration()
+
+    func ifConfigState(_ condition: ExprSyntax) throws -> IfConfigState {
```
Why do we have `CanImportExprSyntax` at all? It seems like we should leave it as a normal call expression and let `SwiftIfConfig` deal with it. There are two diagnostics that would have to move into `SwiftIfConfig` (missing major version, wrong number of arguments), but those are straightforward.

StevenWong12 replied:
That's because the diagnostics(wrong label, wrong number of arguments..) are currently generated in `SwiftParserDiagnostics` and a call expression node is not expressive enough to generate them.

DougGregor replied:
`CanImportExprSyntax` is gone now

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+        }
+
+        let versionText: String
+        if let stringLiteral = secondArg.expression.as(StringLiteralExprSyntax.self),
+          stringLiteral.segments.count == 1,
+          let firstSegment = stringLiteral.segments.first,
+          case .stringSegment(let stringSegment) = firstSegment
+        {
```
Yes, the C++ version of this code (`getCanImportVersion`) allows string literals for the version. Per my other comment, I don't understand why we're parsing `canImport` so completely in the parser, when it can be a normal `CallExprSyntax`.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+      // _underlyingVersion.
+      let version: CanImportVersion
+      if let secondArg = call.argumentList.dropFirst().first {
+        if secondArg.label?.text != "_version" && secondArg.label?.text != "_underlyingVersion" {
```
I'm going to mark this as "resolved", because I think `CanImportExprSyntax` should be removed.

DougGregor commented on Sources/SwiftIfConfig/SyntaxLiteralUtils.swift:
```diff
@@ -0,0 +1,48 @@
+
+extension ExprSyntax {
+  /// Whether this is a simple identifier expression and, if so, what the identifier string is.
+  var simpleIdentifierExpr: String? {
```
I find it easier to work with the string, because that's what clients want. This is in the same vein as the other comment about wanting to have a central "get the identifier for this token" API to use in various places.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+  {
+
+    /// Perform a check for an operation that takes a single identifier argument.
+    func doSingleIdentifierArgumentCheck(_ body: (String, ExprSyntax) -> Bool?) throws -> Bool? {
```
No, because what we're passing in here are functions from the build configuration, and those take `(String, ExprSyntax)`.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+    let fnName = call.calledExpression.simpleIdentifierExpr,
+    let fn = IfConfigFunctions(rawValue: fnName)
+  {
+
```
Fixed

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+  /// used by an older if configuration form `_compiler_version("...")`.
+  fileprivate init(
+    parsingCompilerBuildVersion versionString: String,
+    _ syntax: ExprSyntax
```
Alright, I've updated the name.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+}
+
+/// Evaluate the condition of an `#if`.
+private func evaluateIfConfig(
+  condition: ExprSyntax,
+  configuration: some BuildConfiguration
+) throws -> Bool {
```
Sure

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+      )
+    }
+
+    // If we found a result, return it.
+    if let result { return result }
```
We handle the `nil` case through a common point below, rather than unwrapping each time. It's likely the flow will change here as we think more about handling errors, so I'm not going to fret over it much now.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+        let segment = stringLiteral.segments.first,
+        case .stringSegment(let stringSegment) = segment
+      else {
+        // FIXME: better diagnostic here
```
When better diagnostics infrastructure is in place, sure

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+        let arg = argExpr.simpleIdentifierExpr,
+        let expectedEndianness = Endianness(rawValue: arg)
+      else {
+        result = nil
```
Added a `FIXME`, because it's a good little starter bug.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+        throw IfConfigError.canImportMissingModule(syntax: ExprSyntax(call))
+      }
+
+      let importPath = firstArg.expression.trimmedDescription.split(separator: ".")
```
Yeah, I'll get there. FIXME'd for now.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+  /// insufficient information to make a determination.
+  public init(condition: some ExprSyntaxProtocol, configuration: some BuildConfiguration) throws {
+    // Apply operator folding for !/&&/||.
+    let foldedCondition = try OperatorTable.logicalOperators.foldAll(condition).as(ExprSyntax.self)!
```
Sure to the nicer version, and yes, it probably makes sense to add some `foldAll` overloads that show that we maintain the overall structure.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+  /// If an error occurred while processing any of the `#if` clauses, this function will throw that error.
+  public func activeClause(in configuration: some BuildConfiguration) throws -> IfConfigClauseSyntax? {
+    for clause in clauses {
+      // If there is no condition, we have reached an unconditional clause. Return it.
+      guard let condition = clause.condition else {
+        return clause
+      }
```
Perhaps parser diagnostics should handle it; it's quite clear from the syntax tree that it's a problem. It also shouldn't prevent us from picking an "active" clause (per this library).

DougGregor commented on Sources/SwiftIfConfig/IfConfigRewriter.swift:
````diff
@@ -0,0 +1,277 @@
+/// #endif
+/// ```
+///
+/// the rewriter will eliminate nodes for inactive clauses, leaving only
+/// those nodes that are in active clauses. When rewriting the above given
+/// a build configuration for Linux, the resulting tree will be
````
Right now, it treats them as inactive. As noted earlier, this is a Point For Discussion.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test Windows

DougGregor commented on Sources/SwiftIfConfig/BuildConfiguration.swift:
```diff
@@ -0,0 +1,92 @@
+/// Therefore, many of the queries return `Bool?`, where `nil` indicates
+/// that the answer is not known. Clients that don't have a lot of context
+/// (such as an IDE that does not have access to the compiler command line)
+/// can return `nil`.
```
I went ahead and dropped the optionality from the result type, and instead allow these methods to `throw` if they have a problem.

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,516 @@
+  /// configuration options `DEBUG` and `B` are provided, but `A` is not.
+  public func isActive(in configuration: some BuildConfiguration) throws -> Bool {
+    var currentNode: Syntax = Syntax(self)
+    var currentClause = currentNode.as(IfConfigClauseSyntax.self)
```
Yes, much better! Thank you

DougGregor commented on Sources/SwiftIfConfig/IfConfigRewriter.swift:
```diff
@@ -0,0 +1,277 @@
+    // Determine the active clause within this syntax node.
+    // TODO: Swallows errors
+    guard let activeClause = try? postfixIfConfig.config.activeClause(in: configuration),
+      case .`postfixExpression`(let postfixExpr) = activeClause.elements
```
Nope, I'll remove them.

DougGregor commented on Sources/SwiftIfConfig/IfConfigRewriter.swift:
```diff
@@ -0,0 +1,277 @@
+          continue
+        }
+
+        let innerElements = Syntax(elements).cast(List.self)
```
Yes, good point. We can be more defensive here.

DougGregor commented on Sources/SwiftIfConfig/IfConfigRewriter.swift:
```diff
@@ -0,0 +1,277 @@
+    // Determine the active clause within this syntax node.
+    // TODO: Swallows errors
+    guard let activeClause = try? postfixIfConfig.config.activeClause(in: configuration),
+      case .`postfixExpression`(let postfixExpr) = activeClause.elements
```
No, they aren't. I've removed them.

DougGregor commented:
@swift-ci please test

DougGregor commented:
Alright, major update for this library today, including:
* Introduce complete handling for diagnostics produced via `#if` condition evaluation.
* Add nice test harnesses for all of the APIs.
* Distinguish between "inactive" and "unparsed" regions so we know where to emit syntactic diagnostics.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
I forgot about `PostfixIfConfigExprSyntax`, so I'll have to fix that before this becomes mergable.

DougGregor commented:
Turns out that `PostfixIfConfigExprSyntax` works. I've renamed `IfConfigState` to `ConfiguredRegionState`, which I think I like better. Although it's pretty long.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
More improvements:
* Improve handling of ill-formed code so that we treat an ill-formed condition as "unparsed" consistently and aligns with the compiler
* Cleaned up file organization
* Added support for the new `_hasAtomicBitWidth` in the compiler

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test Windows

DougGregor commented:
@swift-ci please test Windows

ahoppen commented on Sources/SwiftIfConfig/SyntaxProtocol+IfConfig.swift:
```diff
@@ -0,0 +1,95 @@
+    while let parent = currentNode.parent {
+      // If the parent is an `#if` configuration, check whether our current
+      // clause is active. If not, we're in an inactive region. We also
+      // need to determine whether
```
Incomplete comment?

ahoppen commented on Sources/SwiftIfConfig/SyntaxProtocol+IfConfig.swift:
```diff
@@ -0,0 +1,95 @@
+  /// Determine whether the given syntax node is active given a set of
+  /// configured regions as produced by `configuredRegions(in:)`.
+  ///
+  /// This is
+  /// an approximation
```
Incomplete comment?

ahoppen commented on Sources/SwiftIfConfig/SyntaxProtocol+IfConfig.swift:
```diff
@@ -0,0 +1,95 @@
+  ///
+  /// This is
+  /// an approximation
+  public func isActive(
+    inConfiguredRegions regions: [(IfConfigClauseSyntax, ConfiguredRegionState)]
+  ) -> ConfiguredRegionState {
```
Is this intended to be public API? I would expect people to use the `isActive` function defined above.

DougGregor replied:
Yes, it's intended to be public. If you've retrieved the "configured regions" already, this is faster than re-evaluating everything via `isActive(in:)`. I've left a comment to that effect.

ahoppen commented on Sources/SwiftIfConfig/BuildConfiguration.swift:
```diff
@@ -0,0 +1,275 @@
+  /// Any version of the module will suffice.
+  case unversioned
+
+  /// Only modules with the given version or higher will match.
+  case version(VersionTuple)
+
+  /// Only modules where the underlying Clang module has the given version or
+  /// higher will match.
+  case underlyingVersion(VersionTuple)
```
IIRC `canImport` also supports build numbers, like `"20A301"`, so I don’t think we can use a `VersionTuple` here.

https://github.com/swiftlang/swift-syntax/blob/248dcef04d9e03b7fc47905a81fc84c6f6c23837/Tests/SwiftParserTest/translated/IfconfigExprTests.swift#L411-L420

DougGregor replied:
WAT

DougGregor replied:
Oh, these are actually errors in the old parser as well:

#if canImport(A, _version: 20A301) // expected-error {{'A' is not a valid digit in integer literal}}
  let a = 1
#endif

#if canImport(A, _version: "20A301") // expected-error {{cannot parse module version '20A301'}}
  let a = 1
#endif

I've added tests to https://github.com/swiftlang/swift-syntax/pull/2770

ahoppen replied:
Ah, I was slightly misremembering things. What is valid and what I forgot before is that the version number can be inside a string literal. (rdar://121070235)

```swift
canImport(MyLibrary, _version: "1.2.3")
```

DougGregor replied:
It's already supported. There's a test like this:

```
    assertIfConfig(#"canImport(SwiftSyntax, _version: "5.9")"#, .active)
```

ahoppen replied:
Oh, I missed that. Great!

ahoppen commented on Sources/SwiftIfConfig/VersionTuple+Parsing.swift:
```diff
@@ -0,0 +1,105 @@
+
+    /// Record a component after checking its value.
+    func recordComponent(_ value: Int) throws {
+      let limit = components.isEmpty ? 9223371 : 999
```
Where is 9223371 coming from? The value seems pretty arbitrary to me…

DougGregor replied:
Stolen straight out of the Swift compiler code base without thinking ;)

ahoppen commented on Sources/SwiftSyntaxMacrosGenericTestSupport/Assertions.swift:
```diff
@@ -338,9 +338,36 @@ extension DiagnosticSpec {
+  }
+}
+
+public func assertDiagnostic(
```
Did you intentionally make this API public? I would prefer to not expose it. If you just want to use it in `SwiftIfConfigTest` we should mark it as `@_spi(Testing)`.

DougGregor replied:
Yes, I need it. I'll mark it `@_spi`.

ahoppen commented on Tests/SwiftIfConfigTest/Assertions.swift:
```diff
@@ -0,0 +1,153 @@
+  }
+}
+
+/// Assert that the various marked positions in the source code have the
+/// expected active states.
+func assertActiveCode(
+  _ markedSource: String,
+  configuration: some BuildConfiguration = TestingBuildConfiguration(),
+  states: [String: ConfiguredRegionState],
+  file: StaticString = #filePath,
+  line: UInt = #line
+) throws {
```
This is only used in ActiveRegionTests.swift. What do you think about moving it there (and making it `fileprivate`) so that the assertion function is located right next to the tests, which means you don’t need to switch files to see what the assertion does.

I think the same applies to the other assertion functions as well.

DougGregor replied:
Sure.

ahoppen commented on Tests/SwiftIfConfigTest/ActiveRegionTests.swift:
```diff
@@ -0,0 +1,100 @@
+  func testActiveRegions() throws {
+    try assertActiveCode(
+      """
+      4️⃣
+      #if DEBUG
+      0️⃣func f()
+      #elseif ASSERTS
+      1️⃣func g()
+
+      #if compiler(>=8.0)
+      2️⃣func h()
+      #else
+      3️⃣var i
+      #endif
+      #endif
+      5️⃣token
```
Should we rename these markers so that 4 is between 3 and 5 instead of before 0? Also, in the parser tests, we usually start with 1 instead of 0, it would be good to be consistent here as well.

Same in the other tests

DougGregor replied:
This is too nitpicky for me when there's substantive stuff to do.

ahoppen commented on Tests/SwiftIfConfigTest/VisitorTests.swift:
```diff
@@ -0,0 +1,255 @@
+    if let identified = node.asProtocol(NamedDeclSyntax.self) {
+      checkName(name: identified.name.text, node: node)
+    } else if let identPattern = node.as(IdentifierPatternSyntax.self) {
+      // FIXME: Should the above be an IdentifiedDeclSyntax?
```
Are you proposing to change the name of `NamedDeclSyntax` or `IdentifierPatternSyntax`? If so, could you file an issue for that instead of adding a `FIXME` here, which we probably won’t find again.

DougGregor replied:
This is the same sort of issue we're seeing with the lexical name lookup work, where we want some kind of central notion of "syntax node that declares a name"

ahoppen commented on Sources/SwiftIfConfig/SwiftIfConfig.docc/SwiftIfConfig.md:
```diff
@@ -0,0 +1,36 @@
+The `SwiftIfConfig` library provides utilities to determine which syntax nodes are part of a particular build configuration. Each utility requires that one provide a specific build configuration (i.e., an instance of a type that conforms to the <doc:BuildConfiguration> protocol), and provides a different view on essentially the same information:
+
+* <doc:ActiveSyntaxVisitor> and <doc:ActiveSyntaxAnyVisitor> are visitor types that only visit the syntax nodes that are included ("active") for a given build configuration, implicitly skipping any nodes within inactive `#if` clauses.
+* `SyntaxProtocol/removingInactive(in:)` produces a syntax node that removes all inactive regions (and their corresponding `IfConfigDeclSyntax` nodes) from the given syntax tree, returning a new tree that is free of `#if` conditions.
```
Can these be docc links (with two backticks) so that the rendered documentation can link to these methods? 

https://swiftpackageindex.com/swiftlang/swift-syntax/main/documentation/swiftifconfig/swiftifconfig

Same for the other methods below.

DougGregor replied:
DocC is telling me the methods don't exist. I'm not sure how to make this work.

ahoppen replied:
Ah, I think docc might be having issues here because these are extensions. Whatever… 🤷🏽

ahoppen commented on Sources/SwiftIfConfig/SwiftIfConfig.docc/SwiftIfConfig.md:
```diff
@@ -0,0 +1,36 @@
+# `SwiftIfConfig`
```
Looks like titles in docc can have code font. 

https://swiftpackageindex.com/swiftlang/swift-syntax/main/documentation/swiftifconfig/swiftifconfig

DougGregor replied:
*can't. Will fix

ahoppen commented on Release Notes/601.md:
```diff
@@ -11,6 +11,10 @@
+- `Error` protocol now has an `asDiagnostics(at:)` method.
+  - Description: This method translates an error into one or more diagnostics, recognizing `DiagnosticsError` and `DiagnosticMessage` instances or providing its own `Diagnostic` as needed.
+  - Pull Request: https://github.com/swiftlang/swift-syntax/pull/1816
+
```
Could you also add the addition of `SwiftIfConfig` in the release notes?

DougGregor commented on Sources/SwiftIfConfig/ActiveSyntaxRewriter.swift:
```diff
@@ -0,0 +1,308 @@
+  /// considered to have failed and the clauses's elements will be
+  /// removed.
+  public func removingInactive(in configuration: some BuildConfiguration) -> (Syntax, [Diagnostic]) {
+    // First pass: Find all of the active clauses for the #ifs we need to
+    // visit, along with any diagnostics produced along the way. This process
+    // does not change the tree in any way.
+    let visitor = ActiveSyntaxVisitor(viewMode: .sourceAccurate, configuration: configuration)
+    visitor.walk(self)
+
+    // If there were no active clauses to visit, we're done!
+    if visitor.numIfClausesVisited == 0 {
+      return (Syntax(self), visitor.diagnostics)
+    }
+
+    // Second pass: Rewrite the syntax tree by removing the inactive clauses
+    // from each #if (along with the #ifs themselves).
+    let rewriter = ActiveSyntaxRewriter(configuration: configuration)
```
The diagnostics are in terms of the tree before the rewrite occurs, because we will end up removing failed #ifs and therefore the nodes that the diagnostics point at. Hence, one diagnostics pass and one folding pass.

ahoppen replied:
I’m pretty sure that the nodes passed into the `SyntaxRewriter.visit` functions are the ones pre-rewrite. So, I would expect that it should be possible to do it in a single pass. Or am I missing something here?

DougGregor commented on Sources/SwiftIfConfig/IfConfigDecl+IfConfig.swift:
```diff
@@ -0,0 +1,59 @@
+  /// If an error occurrs while processing any of the `#if` clauses,
+  /// that clause will be considered inactive and this operation will
+  /// continue to evaluate later clauses.
+  public func activeClause(
+    in configuration: some BuildConfiguration,
+    diagnosticHandler: ((Diagnostic) -> Void)? = nil
+  ) -> IfConfigClauseSyntax? {
```
I've reworked things to always return diagnostic arrays.

DougGregor commented on Sources/SwiftIfConfig/ActiveSyntaxVisitor.swift:
```diff
@@ -0,0 +1,129 @@
+  public var diagnostics: [Diagnostic] = []
+
+  /// The number of "#if" clauses that were visited.
+  var numIfClausesVisited: Int = 0
```
I'll make this a Bool

DougGregor commented on Sources/SwiftIfConfig/ConfiguredRegionState.swift:
```diff
@@ -0,0 +1,51 @@
+    diagnosticHandler: ((Diagnostic) -> Void)? = nil
+  ) throws {
+    // Apply operator folding for !/&&/||.
+    let foldedCondition = try OperatorTable.logicalOperators.foldAll(condition) { error in
+      diagnosticHandler?(error.asDiagnostic)
+      throw error
```
This is gone based on the earlier comments. It's all [Diagnostic] returns now

DougGregor commented on Sources/SwiftIfConfig/ConfiguredRegionState.swift:
```diff
@@ -0,0 +1,51 @@
+import SwiftSyntax
+
+/// Describes the state of a particular region guarded by `#if` or similar.
+public enum ConfiguredRegionState {
```
I had something similar before. I hate them all, but `IfConfigRegionState` is... I guess?... better than what I have. It's shorter at least.

DougGregor commented on Sources/SwiftIfConfig/ConfiguredRegions.swift:
```diff
@@ -0,0 +1,108 @@
+  var regions: [(IfConfigClauseSyntax, ConfiguredRegionState)] = []
+
+  /// Whether we are currently within an active region.
+  var inActiveRegion = true
```
It's used to distinguish "inactive" vs. "unparsed"

ahoppen replied:
Ah, makes sense. Adding that information as a doc comment in https://github.com/swiftlang/swift-syntax/pull/2762.

DougGregor commented on Sources/SwiftIfConfig/ConfiguredRegions.swift:
```diff
@@ -0,0 +1,108 @@
+        regions.append((clause, isVersioned ? .unparsed : .inactive))
+      }
+
+      // Recurse into inactive (but not unparsed) regions to find any
+      // unparsed regions below.
```
Yeah, the comment on `SyntaxProtocol.configuredRegions` was stale.

DougGregor commented on Sources/SwiftIfConfig/SyntaxLiteralUtils.swift:
```diff
@@ -0,0 +1,47 @@
+      return nil
+    }
+
+    // FIXME: Handle escaping here.
+    return identExpr.baseName.text
```
Hmph, `identifier` returns an optional. We should probably shore that up or introduce a new concept here, because a `DeclReferenceExprSyntax` should always refer to a name.

DougGregor commented:
Addressed most of these comments in https://github.com/swiftlang/swift-syntax/pull/2759

ahoppen commented:
Thanks for addressing the comments. I’m addressing a few more minor ones in https://github.com/swiftlang/swift-syntax/pull/2762. 

The following discussions are still open. Composed this list partially for myself to keep track.

- [ ] Using a single pass in `ActiveSyntaxRewriter`
  - https://github.com/swiftlang/swift-syntax/pull/1816#discussion_r1681126213
- [ ] Emitting diagnostics in case of errors instead of ignoring errors / trapping
  - https://github.com/swiftlang/swift-syntax/pull/1816#discussion_r1681142474
  - https://github.com/swiftlang/swift-syntax/pull/1816#discussion_r1681151598
  - https://github.com/swiftlang/swift-syntax/pull/1816#discussion_r1681155777
- [ ] Renaming `versioned` to eg. `syntaxErrorsAllowed`
  - https://github.com/swiftlang/swift-syntax/pull/1816#discussion_r1681211064
- [ ] Document how `#if` regions are handled if `BuildConfiguration` throws
  - https://github.com/swiftlang/swift-syntax/pull/1816#discussion_r1681274837
- [ ] Consider the use of `Identifier` or a similar concept
  - https://github.com/swiftlang/swift-syntax/pull/1816#discussion_r1693914616 
  - https://github.com/swiftlang/swift-syntax/pull/1816#discussion_r1693879981
- [ ] Support `canImport` with build numbers
  - https://github.com/swiftlang/swift-syntax/pull/1816#discussion_r1681365980
- [ ] Figure out where 9223371 is coming from
  - https://github.com/swiftlang/swift-syntax/pull/1816#discussion_r1681366653

DougGregor commented on Sources/SwiftIfConfig/IfConfigEvaluation.swift:
```diff
@@ -0,0 +1,431 @@
+  condition: ExprSyntax,
+  configuration: some BuildConfiguration,
+  diagnosticHandler: ((Diagnostic) -> Void)?
+) throws -> (active: Bool, versioned: Bool) {
```
https://github.com/swiftlang/swift-syntax/pull/2769

DougGregor commented on Sources/SwiftIfConfig/ConfiguredRegions.swift:
```diff
@@ -0,0 +1,108 @@
+      }
+
+      // For inactive clauses, distinguish between inactive and unparsed.
+      let isVersioned =
+        (try? clause.isVersioned(
+          configuration: configuration,
+          diagnosticHandler: nil
+        )) ?? true
```
Part of https://github.com/swiftlang/swift-syntax/pull/2770
